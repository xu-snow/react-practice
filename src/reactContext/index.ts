import * as  React from "react";

export interface IState {
  color: string;
  theme: string;
  dispatch?: (action: any) => void
}

export const initialState: IState = {
  color: "blue",
  theme: "dark",
};

export const GlobalStoreContext: React.Context<IState> = React.createContext({
  ...initialState
});