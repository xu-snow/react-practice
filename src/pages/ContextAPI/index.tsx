import * as React from 'react';
import { GlobalStoreContext, initialState, IState } from '../../reactContext/index'

type ActionType = 'UPDATE_THEME' | 'UPDATE_COLOR' | 'UPDATE_THEME_THEN_COLOR'
interface IStateDispatch extends IState {
  dispatch: (action: { [key: string]: any, type: ActionType }) => void
}

class GlobalStoreContextProvider extends React.Component {
  // initialState
  public state = {
    ...initialState
  };
  // reducer
  public handleContextChange = (action: { type: ActionType, theme: string, color: string }) => {
    switch (action.type) {
      case "UPDATE_THEME":
        return this.setState({
          theme: action.theme
        });
      case "UPDATE_COLOR":
        return this.setState({
          color: action.color
        });
      case "UPDATE_THEME_THEN_COLOR":
        return this.setState({
          color: action.color,
          theme: action.theme
        });
      default:
        return;
    }
  };

  public render() {
    return (
      <GlobalStoreContext.Provider
        value={{
          color: this.state.color,
          dispatch: this.handleContextChange,
          theme: this.state.theme,
        }}
      >
        {this.props.children}
      </GlobalStoreContext.Provider>
    );
  }
}

const SubComponent = (props: IStateDispatch) => {
  function handleUT() {
    props.dispatch({
      theme: "light",
      type: "UPDATE_THEME",
    })
  }
  function handleUC() {
    props.dispatch({
      color: "red",
      type: "UPDATE_COLOR",
    })
  }
  function handleUTC() {
    props.dispatch({
      color: "purple",
      theme: "dark",
      type: "UPDATE_THEME_THEN_COLOR",
    })
  }
  return (
    <div>
      <button onClick={handleUT}  >
        change theme
      </button>
      <div className="content-theme">{props.theme}</div>
      {/* action */}
      <button onClick={handleUC} >
        change color
      </button>
      <div className="content-color">{props.color}</div>
      {/* action */}
      <button
        className="change-all"
        onClick={handleUTC}
      >
        change theme then color
      </button>
    </div>
  )
};


// tslint:disable-next-line:max-classes-per-file
export default class ContextAPI extends React.Component {
  public render() {
    return (
      <GlobalStoreContextProvider>
        <GlobalStoreContext.Consumer>
          {(context) => (
            <SubComponent
              theme={context.theme}
              color={context.color}
              dispatch={context.dispatch!}
            />
          )}
        </GlobalStoreContext.Consumer>
      </GlobalStoreContextProvider>
    );
  }
}