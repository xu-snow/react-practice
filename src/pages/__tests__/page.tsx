
import {  mount} from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import { initialState, } from '../../reactContext/index'
import ContextAPI from '../ContextAPI';

it('context work fine', () => {
  const wrapper = mount(<ContextAPI />);
  // expect(wrapper.find('div').first().hasClass('content-theme')).toEqual(true)
  expect(wrapper.find('.content-theme').at(0)).toHaveText(initialState.theme)

  wrapper.find('.change-all').first().simulate('click');


  // color: "purple",
  // theme: "dark",
  
  expect(wrapper.find('.content-theme').first()).toHaveText('dark')
  
  expect(wrapper.find('.content-color').first()).toHaveText('purple')

  expect(toJson(wrapper)).toMatchSnapshot();

});