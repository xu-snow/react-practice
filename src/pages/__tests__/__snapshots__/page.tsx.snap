// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`context work fine 1`] = `
<ContextAPI>
  <GlobalStoreContextProvider>
    <SubComponent
      color="purple"
      dispatch={[Function]}
      theme="dark"
    >
      <div>
        <button
          onClick={[Function]}
        >
          change theme
        </button>
        <div
          className="content-theme"
        >
          dark
        </div>
        <button
          onClick={[Function]}
        >
          change color
        </button>
        <div
          className="content-color"
        >
          purple
        </div>
        <button
          className="change-all"
          onClick={[Function]}
        >
          change theme then color
        </button>
      </div>
    </SubComponent>
  </GlobalStoreContextProvider>
</ContextAPI>
`;
